File: python/leetcode/advenced-algorithm/minimum-window-substring.py
from typing import List
from collections import Counter


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need = Counter(t)
        missing = len(t)
        left = start = end = 0
        for right, char in enumerate(s, 1):
            missing -= need[char] > 0
            need[char] -= 1

            if missing == 0:
                print(s[left], need[s[left]])
                while left < right and need[s[left]] < 0:
                    need[s[left]] += 1
                    left += 1
                if not end or right - left <= end - start:
                    start, end = left, right
                    need[s[left]] += 1
                    missing += 1
                    left += 1
        print(start, end)
        return s[start:end]


print(Solution().minWindow("a", "a"))

File: JS&TS/leetcode/advanced-algorithm/minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function (s, t) {
  const need = {};
  let missing = t.length;
  let left = 0;
  let start = 0;
  let end = 0;

  for (let i = 0; i < t.length; i++) {
    if (!need[t[i]]) need[t[i]] = 0;
    need[t[i]]++;
  }

  for (let right = 0; right < s.length; right++) {
    if (!need[s[right]]) need[s[right]] = 0;
    if (need[s[right]] > 0) missing--;
    need[s[right]]--;

    if (missing === 0) {
      while (left < right + 1 && need[s[left]] < 0) {
        need[s[left]]++;
        left++;
      }

      if (!end || right + 1 - left <= end - start) {
        start = left;
        end = right + 1;
        need[s[left]]++;
        missing++;
        left++;
      }
    }
  }

  return s.slice(start, end);
};

console.log(minWindow('a', 'ab'));
